packages/data-broker/src/services/dhis/DhisInputSchemeResolvingApiProxy.js

```diff
@@ -0,0 +1,244 @@
+      // both the dataElements and orgUnits need to be ids
+      modifiedQuery = await this.replaceDataElementCodesWithIds(modifiedQuery);
+      modifiedQuery = await this.replaceOrgUnitCodesWithIds(modifiedQuery);
+      modifiedQuery.inputIdScheme = 'uid';
+    }
+
+    return this.api.getAnalytics(modifiedQuery);
+  }
+
+  async getEventAnalytics(query) {
+    let modifiedQuery = { ...query };
+
+    // The base method here, DhisApi#getEventAnalytics will convert codes to ids before submitting the
+    // actual data request. It does that with separate calls e.g. /api/dataElements. However, with dhis
+    // setups where the Data Elements do not have codes, this api call will give us no new information.
+    // To prevent this, we pre-emptively swap out codes for ids, using our internal mapping, so that
+    // DhisApi#getEventAnalytics does not need to make these code-to-id conversion calls.
+    const { dataElementIdScheme = 'code' } = query;
+
+    const modifyDataElementsToUseDhisId = await this.allDataElementsHaveDhisId(query);
+
+    if (modifyDataElementsToUseDhisId) {
+      modifiedQuery = await this.replaceDataElementCodesWithIds(modifiedQuery);
+      modifiedQuery.dataElementIdScheme = 'id';
+    }
+
+    if (await this.allProgramsHaveDhisId(query)) {
+      modifiedQuery = await this.replaceProgramCodesWithIds(modifiedQuery);
+    }
+
+    if (await this.allOrgUnitsHaveDhisId(query)) {
+      modifiedQuery = await this.replaceOrgUnitCodesWithIds(modifiedQuery);
+    }
+
+    const response = await this.api.getEventAnalytics(modifiedQuery);
+
+    if (dataElementIdScheme !== 'code') {
+      return response;
+    }
+
+    // The api response will contain data elements with ids, and DhisApi will not be able to translate these
+    // back into codes (because the codes are not set in dhis). So, we have to do it ourselves using the internal
+    // mapping.
+    if (!modifyDataElementsToUseDhisId) {
+      return response;
+    }
+
+    const dataElementIdToCode = {};
+
+    const dataElements = await this.models.dataSource.find({ code: query.dataElementCodes });
+
+    for (const dataElement of dataElements) {
+      if (dataElement.config.dhisId) {
+        dataElementIdToCode[dataElement.config.dhisId] = dataElement.code;
+      }
+    }
+
+    return translateElementKeysInEventAnalytics(response, dataElementIdToCode);
+  }
+
+  /**
+   * @param query {*}
+   * @returns bool
+   * @private
+   */
+  allDataElementsHaveDhisId = async query => {
+    const { dataElementCodes } = query;
+
+    const dataElements = await this.models.dataSource.find({ code: dataElementCodes });
+
+    for (const dataElementCode of dataElementCodes) {
+      const dataElement = dataElements.find(d => d.code === dataElementCode);
+
+      if (!dataElement) {
+        // most likely using a data element code that is not defined in data_source, which is supported
+        // by this service
+        return false;
+      }
```

Comment:
@kael89 can also be deleted after your work on https://github.com/beyondessential/tupaia-backlog/issues/663

@IgorNadj a todo comment here too please

Reply:
Added